functions/snippets/src/main/java/HelloHttp.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// [START functions_helloworld_http]
+
+import com.google.cloud.functions.HttpFunction;
+import com.google.cloud.functions.HttpRequest;
+import com.google.cloud.functions.HttpResponse;
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.util.logging.Logger;
+
+public class HelloHttp implements HttpFunction {
+  private static final Logger LOGGER = Logger.getLogger(HelloHttp.class.getName());
+
+  private Gson gsonParser = new Gson();
+
+  @Override
+  public void service(HttpRequest request, HttpResponse response)
+      throws IOException {
+    String name = "world";
+
+    // Parse JSON request and check for "name" field
+    try {
+      JsonElement requestParsed = gsonParser.fromJson(request.getReader(), JsonElement.class);
+      JsonObject requestJson = null;
+
+      if (requestParsed.isJsonObject()) {
+        requestJson = requestParsed.getAsJsonObject();
+      }
+
+      if (requestJson != null && requestJson.has("name")) {
+        name = requestJson.get("name").getAsString();
+      }
+    } catch (JsonParseException e) {
+      LOGGER.severe("Error parsing JSON: " + e.getMessage());
+    }
+
+    if (request.getFirstQueryParameter("name").isPresent()) {
+      name = request.getFirstQueryParameter("name").get();
```

Comment:
I find this a bit confusing. In the previous try/catch, we've gone to all the trouble of parsing a JSON object out of the input, just to extract a `name`, which we then proceed to overwrite here. I think the logic might be a bit clearer if this `if` came first, with an `else` that did the JSON parsing thing.

Reply:
I swapped the ordering of these, but IMO an `else` statement would add an additional layer of indentation/[cyclomatic] complexity.

I don't feel strongly about it though, so I'll add it if you'd prefer me to.

Reply:
I think the goal is to show both how to get query parameters and how to read the body of the request. It would probably be easier to follow if we got different things from the two places. Maybe something like this?

```java
  Optional<String> nameFromBody = Optional.empty();
  try {
    ...
    if (requestJson != null && requestJson.has("name")) {
      nameFromBody = Optional.of(requestJson.get("name").getAsString());
    }
  } catch (...) {
    ...
  }
  Optional<String> nameFromQueryParameter = request.getFirstQueryParameter("name");
  BufferedWriter writer = response.getWriter();
  if (nameFromBody.isPresent()) {
    writer.write(String.format("Name from request body: %s\n", nameFromBody.get());
  }
  if (nameFromQueryParameter.isPresent()) {
    writer.write(String.format("Name from query parameter: %s\n", nameFromQueryParameter.get());
  }
  if (!nameFromBody.isPresent() && !nameFromQueryParameter.isPresent()) {
    writer.write("No name in either request body or query parameter\n");
  }
```